# Les différentes méthodes de tri Python et leur complexité

Sujets :
e1 s42 e2 s43 e1 s45
tri en plus :
fusion et bulle

https://pixees.fr/informatiquelycee/n_site/nsi_prem_tri_algo.html

#### Les algorithmes de tri des éléments d'un tableau ont une place à part en algorithmique, les deux algorithmes de tri les plus classiques sont :  
#### - le tri par insertion,
#### - le tri par sélection.

## 1- Le tri par insertion :

### Fonctionnement :
##### On parcours l'ensemble de la liste et dès qu'un élément "x" est inférieur au précédent, les deux sont échangés et l'on vérifie si le nouvel élément précédent "x" lui soit bien inférieur sinon ils sont encore une fois échangé et ce jusqu'à ce que "x" soit bien inférieur à l'élément suivant, qu'il soit trié. Ensuite le parcours de la liste reprend et applique la méthode vue précédament jusqu'à ce que la liste soit entièrement parcourue.

### Ce shéma explique la méthode utilisée par le tri par insertion :
![](https://pixees.fr/informatiquelycee/n_site/img/nsi_prem_algo_tri_2.png)
### Implémentation du programme de tri par insertion de David Roche :

In [1]:
def tri_par_insertion(tab:list)->list:
    j = 1
    while j < len(tab): # boucle qui parcours l'ensemble de "tab"
        i = j-1 # i l'indice qui commence au rang 0 (= j-1)
        k = tab[j] # k prend l'élément au rang j (= i+1)
        while i>=0 and tab[i] > k: # une seconde boucle qui vérifie que l'élément tab[i] est bien inférieur à k (tab[i+1])
            tab[i+1] = tab[i] # si tab[i] > k (tab[i+1]) on pose tab[i+1] = tab[i]
            i -= 1 # on diminue i de 1 en 1 jusqu'à 0 (arrêt de la boucle while)
        tab[i+1] = k # lorsque i = 0 on remet k dans tab[i+1]
        j += 1 # on avance alors de 1 dans la liste et on répète le processus
    return tab

In [11]:
tab = [27, 10, 12, 8, 11]
tri_par_insertion(tab)

[8, 10, 11, 12, 27]

### Ma propre version du tri par insertion :

In [9]:
def mon_tri_par_insertion(tab:list)->list:
    for parcours in range(len(tab)-1):
        x = 0
        if tab[parcours] > tab[parcours+1]:
            while tab[parcours-x] > tab[parcours-x+1] and parcours-x >= 0:
                tab[parcours-x+1], tab[parcours-x] = tab[parcours-x], tab[parcours-x+1]
                x += 1
    return tab

In [10]:
tab = [27, 10, 12, 8, 11]
mon_tri_par_insertion(tab)

[8, 10, 11, 12, 27]

### Nous avons donc pu voir deux versions du tri par insertion, maintenant nous allons parler de leur complexité :

##### Pour rappel : La complexité algorithmique d’un programme informatique est d’une importance majeure. En effet, c’est elle qui permet de vérifier l’efficacité de ce dernier. Vous l’aurez donc compris, la complexité algorithmique est une grandeur : cela peut être un nombre, mais c’est très souvent un ordre de grandeur.

### Il existe également 2 façon de mesurer la complexité d'un algorithme :
### 1- Avec TimeIT (dans un cas spécifique uniquement) : 

In [15]:
tab = [27, 10, 12, 8, 11]
%timeit tri_par_insertion(tab)

1.32 µs ± 1.88 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


In [16]:
tab = [27, 10, 12, 8, 11]
%timeit mon_tri_par_insertion(tab)

824 ns ± 14.3 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


![](https://images.schoolmouv.fr/4e-maths-c23-img01.png)
#### TimeIT nous donne une grandeur en temps : 1,32 micro seconde contre 824 nano seconde pour mon programme.
### 2- De façon mathématique (plus générale) :
#### Shéma exemple :
![](https://pixees.fr/informatiquelycee/n_site/img/nsi_prem_algo_tri_4.png)
#### Pour un tableau de **longueur 5** nous avons donc **10 déplacements** : 1+2+3+4

#### Résonnement :
#### Pour un tableau de n éléments nous avons donc : 1 + 2 + 3  ...  n-3 + n-2 + n-1 déplacements
#### Ainsi : S = 1 + 2 + 3 ... n-3 + n-2 + n-1
#### Mais en l'inversant on aurait : S' = n-1 + n-2 + n-2  ...  3 + 2 + 1 avec S = S'
#### Or, S + S' = n-1+1 + n-2+2 + n-3+3  ...  n-3+3 + n-2+2 + n-1+1
#### = n + n + n  ...  n + n + n
#### On peut donc en déduire que :
#### S + S' = 2S = n(n-1)
#### Donc S = n(n-1)/2 = n²-n/2 = n²/2 - n/2

## 2- Le tri par selection :

### Fonctionnement :
##### On parcours l'ensemble de la liste pour trouver un minimum et on le déplace au début de la liste puis on la reparcours à partir du deuxième élément (d'indice 1) et on répète le processus.

### Ce shéma explique la méthode utilisée par le tri par selection :
![](https://pixees.fr/informatiquelycee/n_site/img/nsi_prem_algo_tri_6.png)
### Implémentation du programme de tri par selection de David Roche :

In [17]:
def tri_par_selection(tab:list)->list:
    i = 0
    while i < len(tab): # première boucle qui parcours la liste
        j = i+1 # j plus grand que i de 1
        min = i # minimum est i pour le moment
        while j < len(tab): # seconde boucle pour chercher le minimum
            if tab[j] < tab[min]: # si un élément est plus petit que l'ancien minimum,
                min = j # il prend sa place
            j += 1 # on incrément j pour parcourir la liste
        if min != i: # si le minimum à changé,
            tab[i], tab[min] = tab[min], tab[i] # on le déplace au début de la liste
        i += 1 # on incrément i pour parcourir la liste
    return tab

In [18]:
tab = [27, 10, 12, 8, 11]
tri_par_selection(tab)

[8, 10, 11, 12, 27]

### Ma propre version du tri par selection :

In [22]:
def mon_tri_par_selection(tab:list)->list:
    x = 0
    for indice in range(len(tab)):
        minimum = tab[x]
        indice_mini = x
        for j in range (x, len(tab)):
            if tab[j] < minimum:
                minimum = tab[j]
                indice_mini = j
        tab[x], tab[indice_mini] = tab[indice_mini], tab[x]
        x += 1
    return tab

In [23]:
tab = [27, 10, 12, 8, 11]
mon_tri_par_selection(tab)

[8, 10, 11, 12, 27]

### Leur complexité :
### 1- Avec TimeIT :

In [24]:
tab = [27, 10, 12, 8, 11]
%timeit tri_par_selection(tab)

3.05 µs ± 45.7 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [25]:
tab = [27, 10, 12, 8, 11]
%timeit mon_tri_par_selection(tab)

3.5 µs ± 33.1 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


#### On trouve donc 3,05 micro seconde contre 3.5 nano seconde pour mon programme.
### 2- De façon mathématique :
#### Lors de la première boucle, nous avons pour un tableau de longueur 5, **4 comparaisons** : e1 avec e2, e1 avec e3 etc.
#### Puis on répète le processus cette fois en commençant au premier élément : e2 avec e3, e2 avec e4 etc. on a donc **3 comparaison**
#### Et ainsi de suite ce qui nous donne pour un tableau de longueur 5 un total de **10 comparaisons** : 4 + 3 + 2 + 1
#### On remarquera la similarité avec le tri par insertion mais cette fois il s'agit de comparaisons et non de déplacements !
#### Ainsi avec le même résonnement que pour le tri par insertion, on trouve également S = n²/2 - n/2
### Nous avons donc pu voir que le tri par insertion est moins complexe que celui par selection ainsi que leur fonctionnements.

#### Lors de la première boucle, nous avons pour un tableau de longueur 5, **4 comparaisons** : e1 avec e2, e1 avec e3 etc.
#### Puis on répète le processus cette fois en commençant au premier élément : e2 avec e3, e2 avec e4 etc. on a donc **3 comparaison**
#### Et ainsi de suite ce qui nous donne pour un tableau de longueur 5 un total de **10 comparaisons** : 4 + 3 + 2 + 1
#### On remarquera la similarité avec le tri par insertion mais cette fois il s'agit de comparaisons et non de déplacements !
#### Ainsi avec le même résonnement que pour le tri par insertion, on trouve également S = n²/2 - n/2
### Nous avons donc pu voir que le tri par insertion est moins complexe que celui par selection ainsi que leur fonctionnements.

## Il existe également deux autres types de tri :
### - le tri par fusion,
### - le tri par bulle.

## 1- Le tri par fusion :

### Fonctionnement :
##### X

### Ce shéma explique la méthode utilisée par le tri par fusion :
![](https://upload.wikimedia.org/wikipedia/commons/6/60/Mergesort_algorithm_diagram.png)

In [ ]:
def fusion(L1,L2):
    n1 = len(L1)
    n2 = len(L2)
    L12 = [0]*(n1+n2)
    i1 = 0
    i2 = 0
    i = 0
    while i1<n1 and i2<n2:
        if L1[i1] < L2[i2]:
            L12[i] = L1[i1]
            i1 += 1
        else:
            L12[i] = L2[i2]
            i2 += 1
        i += 1
    while i1<n1:
    	L12[i] = L1[i1]
    	i1 += 1
    	i += 1
    while i2<n2:
    	L12[i] = L2[i2]
    	i2 += 1
    	i += 1 
    return L12